invenio_jsonschemas/utils.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio JSONSchemas.
+# Copyright (C) 2015 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Different utils for JSON handling."""
+
+import copy
+import json
+import os.path
+import urlparse
+
+from flask import current_app, url_for
+
+from jsonpointer import resolve_pointer
+
+import jsonschema
+
+from speaklater import _LazyString
+
+
+class InsecureSchemaLocation(Exception):
+
+    """The try to load a JSON schema from an insecure location."""
+
+    pass
+
+
+def urljoin(*args):
+    """Join parts of a URL together, similar to `os.path.join`."""
+    def _add_trailing_slash(s):
+        if s.endswith('/'):
+            return s
+        else:
+            return s + '/'
+
+    if len(args) > 1:
+        result = reduce(urlparse.urljoin, map(_add_trailing_slash, args[:-1]))
+    else:
+        result = ''
+
+    if len(args) > 0:
+        result = urlparse.urljoin(
+            result,
+            args[-1]
+        )
+
+    return result
+
+
+def internal_schema_url(*parts):
+    """Return the URL to an internal JSON schema."""
+    return url_for(
+        'jsonschemas.schema',
+        path=os.path.join(*parts),
+        _external=True
+    )
+
+
+def ensure_subpath(path):
+    """Santize given path to be a subpath, e.g. not traversing up.
+
+    The returned path will not contain any upwards traversal using `..` and
+    therefore can safely appended to a basepath.
+
+    .. caution::
+        This does **NOT** protect against symlink based traversal!
+    """
+    # http://www.guyrutenberg.com/2013/12/06/preventing-directory-traversal-in-python/
+    return os.path.normpath('/' + path).lstrip('/')
```

Comment:
`secure_filename' is designed to generate a filename, not to navigate to a file and ensuring that it is located in a subpath. You proposal would even prevent the placement in ANY subfolder, because it removes all path separators.